sdk/export/metric/metric.go

```diff
@@ -99,35 +106,39 @@ type AggregationSelector interface {
 // MinMaxSumCount aggregator to a Counter instrument.
 type Aggregator interface {
 	// Update receives a new measured value and incorporates it
-	// into the aggregation.  Update() calls may arrive
-	// concurrently as the SDK does not provide synchronization.
+	// into the aggregation.  Update() calls may be called
+	// concurrently.
 	//
 	// Descriptor.NumberKind() should be consulted to determine
 	// whether the provided number is an int64 or float64.
 	//
 	// The Context argument comes from user-level code and could be
-	// inspected for distributed or span context.
+	// inspected for a `correlation.Map` or `trace.SpanContext`.
 	Update(context.Context, metric.Number, *metric.Descriptor) error
 
-	// Checkpoint is called during collection to finish one period
-	// of aggregation by atomically saving the current value.
-	// Checkpoint() is called concurrently with Update().
-	// Checkpoint should reset the current state to the empty
-	// state, in order to begin computing a new delta for the next
-	// collection period.
+	// SynchronizedCopy is called during collection to finish one period
+	// of aggregation by atomically saving the current state into
+	// the argument Aggregator.
```

Comment:
I feel it might be worth adding an explanation about the function signature accepting a `Aggregator`.

In doing so we introduced a new class of error, one based on dynamic typing. I think it is important that we make it clear we chose this trade-off for the performance benefit of reducing the number of allocations (one instead of one-per-call).

Maybe something like this:

```suggestion
	// the argument Aggregator.
	//
	// This method will return an InconsistentAggregatorError
	// if the Aggregator is not the same type as the receiver. This
	// dynamic typing error was introduced for the trade-off of being
	// able to avoid a large number of allocations (as would be the
	// case if the implementation created a new Aggregator to return
	// each call). Given this method is expected have a large number
	// of calls during normal operation, this trade-off seemed appropriate.
```

Reply:
I wrote:

```
	// This method will return an InconsistentAggregatorError if
	// this Aggregator cannot be copied into the destination due
	// to an incompatible type.
```

I am a bit worried that I gave a wrong impression somewhere, about optimizing for allocations. I did change the constructor signatures in this PR to keep the status quo, mainly out of concern for the Accumulator performance, where every distinct new label set allocates two aggregator states.

The major justification I had for this change was that elsewhere in the export pipeline, it is confusing to have to hold two pieces of state when you only need one (and to have to think about whether you've swapped the current state into the checkpoint, as #799 showed).

I will say that the two-state+swap solution (in #799) isn't terrible, but I also see a possibility that user-defined Aggregators will be able to copy into other, compatible types of Aggregator.

With hindsight, I can see ways to improve both #799 and this PR, but I'd like to merge. Those improvements would have been / would be:

1. In #799, I only needed to add Swap, the two accessors weren't needed. The Integrator could just use Swap before and after calling Exporters, so that Exporters always access what is in the checkpointed "slot".
2. In this PR, instead of changing the Aggregator constructors to return an array of initialized _package_.Aggregators, it could have allowed the AggregatorSelector to allocate a slice of uninitialized _package_.Aggregators and called a `Reset()` or `Zero()` method on them, whereas currently you must use `Checkpoint()` to zero the Aggregator state.

nit: I also realized that `SynchronizedCopy` implementations are not testing that Aggregators have similar configurations. For example, MMSC has a `kind` field that probably should be copied. I will merge this and continue improving, there's something still not right about several things here. (I left a TODO in the code.)

I would be willing to remove the dynamic type checks. The `AggregatorFor()` API already states that the returned Aggregator has to be consistent so that Merge works. It MUST give Aggregators so that all pairs are compatible, I would say.